odoo/tools/translate.py

```diff
@@ -443,155 +444,216 @@ def translate_sql_constraint(cr, key, lang):
+        try:
+            from odoo.http import request  # noqa: PLC0415
+            request_env = request.env
+            if request_env and (lang := request_env.lang):
+                return lang
+        except RuntimeError:
+            pass
+        # Last resort: attempt to guess the language of the user
+        # Pitfall: some operations are performed in sudo mode, and we
+        #          don't know the original uid, so the language may
+        #          be wrong when the admin language differs.
+        cr = self._get_cr(frame)
+        uid = self._get_uid(frame)
+        if cr and uid:
+            env = odoo.api.Environment(cr, uid, {})
+            if lang := env['res.users'].context_get().get('lang'):
+                return lang
+        # fallback
+        if default_lang:
+            _logger.debug('no translation language detected, fallback to %s', default_lang)
+            return default_lang
+        # give up
+        _logger.log(log_level, 'no translation language detected, skipping translation %s', frame, stack_info=True)
+        return ''
 
     def __call__(self, source, *args, **kwargs):
-        translation = self._get_translation(source)
         assert not (args and kwargs)
-        if args or kwargs:
-            if any(isinstance(a, Markup) for a in itertools.chain(args, kwargs.values())):
-                translation = escape(translation)
-            try:
-                return translation % (args or kwargs)
-            except (TypeError, ValueError, KeyError):
-                bad = translation
-                # fallback: apply to source before logging exception (in case source fails)
-                translation = source % (args or kwargs)
-                _logger.exception('Bad translation %r for string %r', bad, source)
-        return translation
-
-    def _get_translation(self, source, module=None):
-        try:
-            frame = inspect.currentframe().f_back.f_back
-            lang = self._get_lang(frame)
-            if lang and lang != 'en_US':
-                if not module:
-                    path = inspect.getfile(frame)
-                    path_info = odoo.modules.get_resource_from_path(path)
-                    module = path_info[0] if path_info else 'base'
-                return code_translations.get_python_translations(module, lang).get(source, source)
-            else:
-                _logger.debug('no translation language detected, skipping translation for "%r" ', source)
-        except Exception:
-            _logger.debug('translation went wrong for "%r", skipped', source)
-                # if so, double-check the root/base translations filenames
-        return source
+        assert isinstance(source, str)
+        module, lang = self._get_translation_source(1)
+        return get_translation(module, lang, source, args or kwargs)
+
+    def _get_translation_source(self, stack_level: int, module: str = '', lang: str = '', default_lang: str = '') -> tuple[str, str]:
```

Comment:
I have the feeling that this method and other associated ones could be static functions used by both `_` and `_lt`.

Reply:
`GettextAlias` has no state, so even the class could be removed. I think it was like this to avoid adding more functions to the main namespace. I will simplify it.